File: api/zodiac.py
from datetime import datetime
from typing import Dict, Any


def get_zodiac_sign(birth_date: str) -> Dict[str, Any]:
    """Calculate zodiac sign and its characteristics from birth date."""
    date = datetime.strptime(birth_date, "%Y-%m-%d")
    month = date.month
    day = date.day

    # Basic zodiac sign data - recommendations will come from AI
    if (month == 3 and day >= 21) or (month == 4 and day <= 19):
        return {
            "name": "Aries",
            "symbol": "♈",
            "element": "Fire",
            "date_range": "March 21 - April 19",
        }
    elif (month == 4 and day >= 20) or (month == 5 and day <= 20):
        return {
            "name": "Taurus",
            "symbol": "♉",
            "element": "Earth",
            "date_range": "April 20 - May 20",
        }
    elif (month == 5 and day >= 21) or (month == 6 and day <= 20):
        return {
            "name": "Gemini",
            "symbol": "♊",
            "element": "Air",
            "date_range": "May 21 - June 20",
        }
    elif (month == 6 and day >= 21) or (month == 7 and day <= 22):
        return {
            "name": "Cancer",
            "symbol": "♋",
            "element": "Water",
            "date_range": "June 21 - July 22",
        }
    elif (month == 7 and day >= 23) or (month == 8 and day <= 22):
        return {
            "name": "Leo",
            "symbol": "♌",
            "element": "Fire",
            "date_range": "July 23 - August 22",
        }
    elif (month == 8 and day >= 23) or (month == 9 and day <= 22):
        return {
            "name": "Virgo",
            "symbol": "♍",
            "element": "Earth",
            "date_range": "August 23 - September 22",
        }
    elif (month == 9 and day >= 23) or (month == 10 and day <= 22):
        return {
            "name": "Libra",
            "symbol": "♎",
            "element": "Air",
            "date_range": "September 23 - October 22",
        }
    elif (month == 10 and day >= 23) or (month == 11 and day <= 21):
        return {
            "name": "Scorpio",
            "symbol": "♏",
            "element": "Water",
            "date_range": "October 23 - November 21",
        }
    elif (month == 11 and day >= 22) or (month == 12 and day <= 21):
        return {
            "name": "Sagittarius",
            "symbol": "♐",
            "element": "Fire",
            "date_range": "November 22 - December 21",
        }
    elif (month == 12 and day >= 22) or (month == 1 and day <= 19):
        return {
            "name": "Capricorn",
            "symbol": "♑",
            "element": "Earth",
            "date_range": "December 22 - January 19",
        }
    elif (month == 1 and day >= 20) or (month == 2 and day <= 18):
        return {
            "name": "Aquarius",
            "symbol": "♒",
            "element": "Air",
            "date_range": "January 20 - February 18",
        }
    else:  # Pisces (February 19 - March 20)
        return {
            "name": "Pisces",
            "symbol": "♓",
            "element": "Water",
            "date_range": "February 19 - March 20",
        }

File: apps/public-app/src/lib/utils/zodiac.ts
interface ZodiacSign {
	name: string;
	symbol: string;
	element: string;
	dateRange: string;
	traits: string[];
	favorableActivities: string[];
}

export const zodiacSigns: Record<string, ZodiacSign> = {
	aries: {
		name: 'Aries',
		symbol: '♈',
		element: 'Fire',
		dateRange: 'March 21 - April 19',
		traits: ['Leadership', 'Energy', 'Adventure'],
		favorableActivities: ['Starting new projects', 'Physical activities', 'Competitive events']
	},
	taurus: {
		name: 'Taurus',
		symbol: '♉',
		element: 'Earth',
		dateRange: 'April 20 - May 20',
		traits: ['Stability', 'Patience', 'Determination'],
		favorableActivities: ['Financial planning', 'Creative work', 'Nature activities']
	},
	gemini: {
		name: 'Gemini',
		symbol: '♊',
		element: 'Air',
		dateRange: 'May 21 - June 20',
		traits: ['Adaptability', 'Communication', 'Curiosity'],
		favorableActivities: ['Learning', 'Social events', 'Writing']
	},
	cancer: {
		name: 'Cancer',
		symbol: '♋',
		element: 'Water',
		dateRange: 'June 21 - July 22',
		traits: ['Intuition', 'Nurturing', 'Emotional depth'],
		favorableActivities: ['Family gatherings', 'Home improvement', 'Self-care']
	},
	leo: {
		name: 'Leo',
		symbol: '♌',
		element: 'Fire',
		dateRange: 'July 23 - August 22',
		traits: ['Confidence', 'Creativity', 'Generosity'],
		favorableActivities: ['Performance', 'Leadership roles', 'Creative expression']
	},
	virgo: {
		name: 'Virgo',
		symbol: '♍',
		element: 'Earth',
		dateRange: 'August 23 - September 22',
		traits: ['Analysis', 'Practicality', 'Service'],
		favorableActivities: ['Organization', 'Health routines', 'Problem-solving']
	},
	libra: {
		name: 'Libra',
		symbol: '♎',
		element: 'Air',
		dateRange: 'September 23 - October 22',
		traits: ['Balance', 'Harmony', 'Justice'],
		favorableActivities: ['Negotiations', 'Art appreciation', 'Social harmony']
	},
	scorpio: {
		name: 'Scorpio',
		symbol: '♏',
		element: 'Water',
		dateRange: 'October 23 - November 21',
		traits: ['Intensity', 'Passion', 'Investigation'],
		favorableActivities: ['Research', 'Transformation', 'Deep connections']
	},
	sagittarius: {
		name: 'Sagittarius',
		symbol: '♐',
		element: 'Fire',
		dateRange: 'November 22 - December 21',
		traits: ['Optimism', 'Adventure', 'Philosophy'],
		favorableActivities: ['Travel', 'Learning', 'Sports']
	},
	capricorn: {
		name: 'Capricorn',
		symbol: '♑',
		element: 'Earth',
		dateRange: 'December 22 - January 19',
		traits: ['Ambition', 'Discipline', 'Responsibility'],
		favorableActivities: ['Career planning', 'Goal setting', 'Business']
	},
	aquarius: {
		name: 'Aquarius',
		symbol: '♒',
		element: 'Air',
		dateRange: 'January 20 - February 18',
		traits: ['Innovation', 'Humanitarianism', 'Independence'],
		favorableActivities: ['Networking', 'Technology', 'Social causes']
	},
	pisces: {
		name: 'Pisces',
		symbol: '♓',
		element: 'Water',
		dateRange: 'February 19 - March 20',
		traits: ['Intuition', 'Creativity', 'Compassion'],
		favorableActivities: ['Artistic pursuits', 'Meditation', 'Helping others']
	}
};

export function getZodiacSign(birthDate: string): ZodiacSign {
	const date = new Date(birthDate);
	const month = date.getMonth() + 1;
	const day = date.getDate();
	const mmdd = `${month.toString().padStart(2, '0')}-${day.toString().padStart(2, '0')}`;

	if (mmdd >= '03-21' && mmdd <= '04-19') return zodiacSigns.aries;
	if (mmdd >= '04-20' && mmdd <= '05-20') return zodiacSigns.taurus;
	if (mmdd >= '05-21' && mmdd <= '06-20') return zodiacSigns.gemini;
	if (mmdd >= '06-21' && mmdd <= '07-22') return zodiacSigns.cancer;
	if (mmdd >= '07-23' && mmdd <= '08-22') return zodiacSigns.leo;
	if (mmdd >= '08-23' && mmdd <= '09-22') return zodiacSigns.virgo;
	if (mmdd >= '09-23' && mmdd <= '10-22') return zodiacSigns.libra;
	if (mmdd >= '10-23' && mmdd <= '11-21') return zodiacSigns.scorpio;
	if (mmdd >= '11-22' && mmdd <= '12-21') return zodiacSigns.sagittarius;
	if (mmdd >= '12-22' || mmdd <= '01-19') return zodiacSigns.capricorn;
	if (mmdd >= '01-20' && mmdd <= '02-18') return zodiacSigns.aquarius;
	return zodiacSigns.pisces;
}
